neo/Consensus/ConsensusService.cs

```diff
@@ -268,7 +268,16 @@ private void OnConsensusPayload(ConsensusPayload payload)
         private void OnPersistCompleted(Block block)
         {
             Log($"persist block: {block.Hash}");
-            block_received_time = TimeProvider.Current.UtcNow;
+            reference_block_time = TimeProvider.Current.UtcNow;
+            uint theoreticalDelay = reference_block_time.ToTimestamp() - block.Timestamp;
+            // maximum expected delay is 5 seconds (this can be moved as a configuration parameter along with block time
+            uint maxDelayToAdvance = 5;
+            theoreticalDelay = theoreticalDelay > maxDelayToAdvance ? maxDelayToAdvance : theoreticalDelay;
+            // unsigned int does not need the need the next line, right? TODO
+            theoreticalDelay = theoreticalDelay < 0 ? 0 : theoreticalDelay;
+            Log($"(long)-theoreticalDelay:{(long)-theoreticalDelay}");
+            reference_block_time.AddSeconds((long)-theoreticalDelay);
```

Comment:
There may be differences in system time between consensus nodes, and the error is likely to reach a dozen seconds. So this adjustment may not help the consensus time.

Reply:
We tested with delays and timestamps not synced and it really helped, Erik. It does not solve the problem, but it is an easy and elegant formula that may help us at this moment with complex changes.

There are no side effects of this PR, we are still using local timestamps for writing blocks, this is just a minor **limited** adjustment on latency occurred during the 3~4 communications needed for creating last block.

As we may noticed, currently, there are rare blocks that are less than 15s. 
This PR will not change this, because the `theoreticalDelay` only advances `PrepareRequest` when block is delayed compared to UTC time, which will be normally used for time-stamping.

Reply:
Which limit do you recommend for `maxDelayToAdvance`, @erikzhang?
Jeff suggested it to be 1/3 of blocktime, but I still prefer static based on values found in the mainnet.

Reply:
> There may be differences in system time between consensus nodes, and the error is likely to reach a dozen seconds. So this adjustment may not help the consensus time.

@erikzhang how is it possible? system clocks don't naturally get differences up to dozen seconds, unless something is very wrong with the system.  Timed protocols like this also depend on having a precise timer, so if we don't have that, it needs to be fixed too.

Reply:
This is very common in winter, if there is no automatic timing.

Reply:
Summerrrr time!! :dagger: Good  time for you to come to Brazil..ahgaahuea

But this a case in which would generate a block with +1s (and just advance with `maxDelayToAdvance`), as limited here:

https://github.com/neo-project/neo/blob/9831f309762ca6eedd49f02ddb3cc892287d9e2f/neo/Consensus/ConsensusContext.cs#L350

For the others cases this PR would help.


PS: In fact, that line should also have a Min Max for solving that winter or summer case.

Reply:
> This is very common in winter, if there is no automatic timing.

I see that timing drifts depending on temperature, several seconds a day. I'm just saying that automatic timing should be necessary for this kind of algorithms to work well... perhaps it gives security some issue, to have a trusted source of time, but I don't see any way around that. Anyway, it's outside of the scope of this discussion here. 
@vncoelho without a precise timing, there's little we could change now, besides improving primary changeview timeout.

Reply:
I got, brother.

However, this formula does not change the way we timeout, it is just an error correction, it has no side effects that I detected, just benefits.
But maybe I am missing some case.

Reply:
I will change that line in another PR and then we keep this discussion here.

Reply:
Created #627

Reply:
> This is very common in winter, if there is no automatic timing.

The CN nodes are not running outside in harsh environments though are they? Also, if CN operators were generally supposed to use a source of time that is synced to an atomic clock, shouldn’t this be the goal. Syncing by using agreement between separately controlled reputable NTP servers that represent atomic clock time should be sufficient, right?